doc/reference/trd-non-xip.md

```diff
@@ -0,0 +1,94 @@
+
+This document provides information relating to running Tock on platforms without
+execute-in-place flash storage.
+
+Introduction
+===============================
+
+Traditional embedded systems use execute-in-place (XIP) flash to both store
+program code and to enable executing that program code. However, newer hardware
+designs, particularly those integrated into a system-on-chip alongside an application processor, do not include 
+executable persistent storage. These platforms, subsequently referred to as "non-xip", can only
+execute instructions stored in volatile RAM. All executable code must be stored
+in external nonvolatile storage, loaded into RAM, and then executed.
+
+This presents challenges for executing Tock on such platforms as Tock generally
+expects to not only have XIP flash, but enough flash to support the overhead of
+discrete applications and their headers. Concretely, Tock platforms generally
+have at least 512 kB of flash. Supporting non-xip platforms may require new
+interfaces or introduce design considerations that Tock on traditional platforms
+would otherwise not consider.
+
+1.0 Background
+===============================
+
+A non-xip platform has no persistent flash storage where compiled binaries can
+be stored and executed. All static program code and data must be stored in some
+nonvolatile storage and then be loaded into RAM to execute.
+
+This system architecture shrinks the amount of the hardware that must be trusted
+when securely executing programs. By not including execute-in-place flash, the
+flash storage does not need to be trusted to correctly store program code and
+data. When the program is copied to RAM the program binary can be verified
+(i.e., by checking a cryptographic signature) before the loaded binary is
+allowed to execute.
+
+
+
+2.0 Tradeoffs
+===============================
+
+Non-xip platforms introduce a new design point for embedded devices with a
+different set of tradeoffs compared to traditional platforms. To help motivate
+potential Tock designs for non-xip platforms, we discuss both the limitations
+compared to traditional platforms, and the properties that may be important on
+traditional platforms but are likely not a priority on non-xip platforms.
+
+Limitations:
+
+- Non-xip platforms have less RAM that traditional platforms have XIP flash.
+  Further, this RAM on non-xip platforms must (at runtime) store both the
+  executable code and the conventional memory items (e.g., the stack and heap).
+  This results in significantly less room for program code than on traditional
+  platforms.
+- As flash storage may not be trustworthy, code retrieved from flash storage may need to be
+  verified before it is used.
+
+Non-limitations:
+
+- Energy is not a bottleneck, and additional power draw from having to load
+  program code into RAM is acceptable.
+- Execution time has relaxed deadlines, and additional execution time from
+  having to load program code into RAM is acceptable.
```

Comment:
Does this need to be qualified on "For some applications"?

Reply:
These non-xip platforms have to give up _something_ right? They can't possibly be _as_ low power and _as_ fast with 25%-50% of the code storage as a conventional MCU can they? I can't imagine we can develop something if we can't assume some tradeoffs in all non-xip cases.

Reply:
Yeah, I think what I was probing at here was that _some_ (and likely few) applications could be time sensitive and would need to perhaps be 'pinned' or similar, which is another design point to consider.

Reply:
On many of these platforms program code is only loaded into RAM after reset, so that is the only time there is an execution delay. Otherwise, these platforms often run faster than XIP platforms, as SRAM accesses are generally faster than NOR flash accesses.

The need to "page in" applications from flash at runtime is definitely not universal to all non-XIP MCU platforms -- just those with significantly less flash than SRAM.

Reply:
Yes, though @hudson-ayers, two specific use cases that came for discussions are those in which there actually is not enough SRAM to concurrently have loaded all the processes that one might want. E.g., TPM2 is just too darn big to fit with more than one or two other processes on a reasonably sized platform.

In one use case, swapping was mentioned as an option, in another, just killing apps on eviction would be acceptable, but both cases involve loading/unloading from SRAM at runtime.

And in both of these cases, I think _some_ applications would prefer to be pinned.

Basically both sides seem to be true here.

Reply:
Dynamically loading parts of apps is definitely an important use case. Tock [already supports that](https://github.com/tock/tock/blob/master/kernel/src/platform/platform.rs#L262) :)